### PR TITLE
store: various small near_store::db::refcount changes

### PR DESCRIPTION
**Pull request by mina86**
Firstly, change decode_value_with_rc so that it returns (None, 0) if
reference count is zero.  This is what the documentation claims it
should do but in reality the function returned (Some(...), 0).

Secondly, change get_with_rc_logic such that it no longer creates
a new Vec.  Instead, Vec which is passed to it is truncated when
necessary.  This reduces amount of times allocator is called.

Thirdly, change refcount_merge so that it first goes through all the
values and only at the end performs single Vec allocation.  This makes
merge_refcounted_records function unnecessary and simplifies code
a bunch.

### CODE REVIEW

File: core/store/src/db.rs [hunk 1]
```diff
@@ -432,12 +431,17 @@ impl Database for TestDB {
...
                 DBOp::UpdateRefcount { col, key, value } => {
-                    let mut val = db[col].get(&key).cloned().unwrap_or_default();
-                    merge_refcounted_records(&mut val, &value);
-                    if !val.is_empty() {
-                        db[col].insert(key, val);
-                    } else {
+                    let existing = db[col].get(&key).map(Vec::as_slice);
+                    let operands = std::iter::once(value.as_slice());
+                    let merged = refcount::refcount_merge(existing, operands);
+                    if merged.is_empty() {
```
Comment (matklad) [inline core/store/src/db.rs:437]: :) I have `if/elses` with negated conditions, thanks!

File: core/store/src/db/refcount.rs [hunk 2]
```diff
@@ -137,7 +101,10 @@ fn refcount_merge<'a>(
 /// values are treated as values with reference count zero.
 pub(crate) fn get_with_rc_logic(column: DBCol, value: Option<Vec<u8>>) -> Option<Vec<u8>> {
     if column.is_rc() {
-        value.and_then(|vec| decode_value_with_rc(&vec).0.map(|v| v.to_vec()))
+        let mut value = value?;
+        decode_value_with_rc(&value).0?;
+        value.truncate(value.len() - 8);
```
Comment (matklad) [inline core/store/src/db/refcount.rs:106]: Took me some thinking to realize why this is correct even if we get an empty value due to merge: I missed the previous line with side-effectful `.0?`. Maybe expand it into explicit match? Can debug-assert that rc is possitive as well then.

File: core/store/src/db/refcount.rs [hunk 3]
```diff
@@ -161,15 +128,14 @@ where
...
-    /// DBCol::State has refcounted values.
     /// Merge adds refcounts, zero refcount becomes empty value.
     /// Empty values get filtered by get methods, and removed by compaction.
     pub(crate) fn refcount_merge(
         _new_key: &[u8],
         existing: Option<&[u8]>,
         operands: &rocksdb::MergeOperands,
     ) -> Option<Vec<u8>> {
-        self::refcount_merge(existing, operands)
+        Some(self::refcount_merge(existing, operands))
```
Comment (matklad) [inline core/store/src/db/refcount.rs:138]: Makes me wonder: should we return `None` in some cases? Would that remove the value as I'd expect? It seems like that could be potentially important, to gc keys?

Comment (matklad) [reply core/store/src/db/refcount.rs:138]: Also perhaps worth filing separately, as that would change the behavior in, *hopefully*, non-observable way.

**Review by matklad — APPROVED** [review]
Great cleanup, thanks!

File: core/store/src/db/refcount.rs [hunk 4]
```diff
@@ -113,18 +70,25 @@ pub(crate) fn encode_negative_refcount(rc: std::num::NonZeroU32) -> Vec<u8> {
 /// Assumes that all provided values with positive reference count have the same
 /// value so that the function is free to pick any of the values.  In build with
 /// debug assertions panics if this is not true.
-fn refcount_merge<'a>(
+pub(crate) fn refcount_merge<'a>(
     existing: Option<&'a [u8]>,
     operands: impl std::iter::IntoIterator<Item = &'a [u8]>,
```
Comment (matklad) [inline core/store/src/db/refcount.rs:75]:
```suggestion
    operands: IntoIterator<Item = &'a [u8]>,
```

It's in prelude

File: core/store/src/db/refcount.rs (hunk 3 above)
Comment (mina86) [reply core/store/src/db/refcount.rs:138]: I did look into that but got failures.  I’m not sure if returning None from merge operator does what we think it does. ;)  I want to take closer look at it separately.

**Pull request merged**